client/src/main/java/io/pravega/client/tables/impl/TableEntryHelper.java

```diff
@@ -0,0 +1,147 @@
+        this.selector = selector;
+        this.config = config;
+        this.totalKeyLength = this.config.getTotalKeyLength();
+    }
+
+    TableSegmentKey toTableSegmentKey(TableSegment tableSegment, Remove removal) {
+        validateKeyVersionSegment(tableSegment, removal.getVersion());
+        return toTableSegmentKey(serializeKey(removal.getKey()), removal.getVersion());
+    }
+
+    private TableSegmentKey toTableSegmentKey(ByteBuf key, Version keyVersion) {
+        return new TableSegmentKey(key, toTableSegmentVersion(keyVersion));
+    }
+
+    TableKey fromTableSegmentKey(TableSegmentKey key) {
+        DeserializedKey rawKey = deserializeKey(key.getKey());
+        return new TableKey(rawKey.primaryKey, rawKey.secondaryKey);
+    }
+
+    TableSegmentEntry toTableSegmentEntry(TableSegment tableSegment, TableEntryUpdate update) {
+        TableKey key = update.getKey();
+        if (update instanceof Put) {
+            validateKeyVersionSegment(tableSegment, update.getVersion());
+        }
+        return new TableSegmentEntry(toTableSegmentKey(serializeKey(key), update.getVersion()), serializeValue(update.getValue()));
+    }
+
+    TableEntry fromTableSegmentEntry(TableSegment s, TableSegmentEntry e) {
+        if (e == null) {
+            return null;
+        }
+
+        Version version = new VersionImpl(s.getSegmentId(), e.getKey().getVersion());
+        TableKey key = fromTableSegmentKey(e.getKey());
+        ByteBuffer value = deserializeValue(e.getValue());
+        return new TableEntry(key, version, value);
+    }
+
+    private TableSegmentKeyVersion toTableSegmentVersion(Version version) {
+        return version == null ? TableSegmentKeyVersion.NO_VERSION : TableSegmentKeyVersion.from(version.asImpl().getSegmentVersion());
+    }
+
+
+    ByteBuf serializeKey(TableKey k) {
+        return serializeKey(k.getPrimaryKey(), k.getSecondaryKey());
+    }
+
+    ByteBuf serializeKey(ByteBuffer primaryKey, ByteBuffer secondaryKey) {
+        Preconditions.checkArgument(primaryKey.remaining() == this.config.getPrimaryKeyLength(),
+                "Invalid Primary Key Length. Expected %s, actual %s.", this.config.getPrimaryKeyLength(), primaryKey.remaining());
+        if (this.config.getSecondaryKeyLength() == 0) {
+            Preconditions.checkArgument(secondaryKey == null || secondaryKey.remaining() == this.config.getSecondaryKeyLength(),
+                    "Not expecting a Secondary Key.");
+            return Unpooled.wrappedBuffer(primaryKey);
+        } else {
+            Preconditions.checkArgument(secondaryKey.remaining() == this.config.getSecondaryKeyLength(),
+                    "Invalid Secondary Key Length. Expected %s, actual %s.", this.config.getSecondaryKeyLength(), secondaryKey.remaining());
+            return Unpooled.wrappedBuffer(primaryKey, secondaryKey);
+        }
+    }
+
+    private DeserializedKey deserializeKey(ByteBuf keySerialization) {
```

Comment:
Can this be ByteBuffer?

Reply:
The `TableSegment` API uses ByteBuf in all its APIs. This is preexisting code and I didn't add it in this feature branch.

This is the same reason why `SegmentIteratorArgs` has `ByteBuf` instead of `ByteBuffer`. All Table Segment APIs use that, while externally we expose `ByteBuffer`. 

I get it that you're trying to get rid of ByteBuf in the Client and I believe that is a worthwhile effort. Would you be OK if I create a follow-up issue to be done after this PR (or after we merge the feature branch in)? Changing ByteBuf to ByteBuffer requires delicate code changes to ensure we don't flip the read/write state accidentally or do other things that can corrupt the data. I prefer to have a dedicated PR that will cleanse it. The good part about it is that it won't touch the external API - it's all internals so from the POV of the user, nothing should change.

Reply:
That's fine